### ISSUE TITLE
Strange type error on `attrs` field with converter (affected by file order) since 1.1.384

### ISSUE THREAD

**Issue by Azureblade3808** [microsoft/pyright#9252]
Look at following 4 files -

a.py
```python
import attrs

@attrs.mutable(slots=False)
class A:
    x: str = attrs.field(converter=str)
```

b.py
```python
import attrs

from .a import A

@attrs.mutable
class B(A):
    pass
```

c.py
```python
import attrs

from .d import D

@attrs.mutable
class C(D):
    pass
```

d.py
```python
import attrs

@attrs.mutable(slots=False)
class D:
    x: str = attrs.field(converter=str)
```

We can see that "a.py" is similar to "d.py" and "b.py" is similar to "c.py".

When they get placed into one folder, Pyright (1.1.384 and 1.1.385, both Pylance and CLI) complains about `D.x` (but not `A.x`).

![Image](https://github.com/user-attachments/assets/b0390019-6e48-4578-b6d2-c9b6b6a7e253)

The only significant difference between `A` and `D` is the file order, and I believe that's the key to this issue.

**Comment by erictraut** [comment]
There are several issues here.

First, the "override" error is being reported in the wrong location.

Second, the error probably shouldn't be reported in the first place. One could argue this is an invalid override, but it's not unsafe, and it's not something that a user has any control over. It's due to the fact that attrs synthesizes a new descriptor object on the subclass for fields that are found in the parent class.

Third, if an error is generated, it should be consistent and not dependent on file import ordering.

@debonte, this regression is a result of the recent change you made to [add a declaration](https://github.com/microsoft/pyright/commit/f55ee0fe8adf235ae53b5215c5d0c5ca3648fced) for the synthesized converter descriptor object. The problem is that there are cases where there is no valid declaration, such as when a new converter descriptor is synthesized in a subclass (such as `C` in the example above) based on the presence of a field in a parent class (`D` in the example above). Modeling this as a new symbol in the child class is correct, but this symbol should not have a declaration in this case. We may need to come up with some other way to provide semantic token types for fields with converters. Let me know if you have any ideas.

**Comment by debonte** [comment]
> We may need to come up with some other way to provide semantic token types for fields with converters. Let me know if you have any ideas.

FWIW, semantic tokenization was just the symptom reported by the user. The missing declaration also preventing Go to Def from working on usages of the converted field. There may be other issues as well.

**Comment by erictraut** [comment]
> The missing declaration also preventing Go to Def from working on usages of the converted field.

The problem is that there is no explicit declaration in the code for the converter descriptor object, so modeling this as a declaration is going to inevitably cause problems. The converter object is created by behind-the-scenes magic within attrs when a "converter" argument is passed to it, and it's recreated for every subclass. This is why the type evaluator code previously didn't provide a declaration in this case. I was nervous when you added the declaration because I suspected it would cause problems, but I failed to come up with any specific examples.

I can understand why your instinct was to add a declaration. It allows the language services modules do the expected thing in most cases, but it is a kludge and has other bad side effects elsewhere in the type evaluator.

In the example above, the internal declaration synthesized for `C.x` indicates that the declaration comes from module `c`, but the parser node associated with the same declaration is located in module `d`. That's clearly wrong, as the associated parse node should be within the module where the declaration is found.

Thinking through our options... Perhaps we should revert the change that synthesizes a pseudo-declaration and enhance the "synthesized type" mechanism in the Symbol class so it can track a "go to" node. This would give the logic in the definition provider enough information to take the user to the correct location when they choose "Go To Definition", and it would avoid confusing the rest of the type checker logic. 

Any other suggestions?

**Comment by debonte** [comment]
That sounds reasonable. I don't have any other suggestions.

Btw, I based my implementation off of https://github.com/microsoft/pyright/blob/main/packages/pyright-internal/src/analyzer/namedTuples.ts#L221. And I see `Symbol.addDeclaration` calls in `typedDicts.ts` as well. So those may need to be fixed also.

**Comment by erictraut** [comment]
@debonte, I've introduced a fix in the type evaluator and made corresponding improvements in the hover and definition providers. I don't have visibility into pylance's semantic token provider. It may also require some updates.

**Comment by debonte** [comment]
@erictraut, I noticed that with your fix a docstring on a converted dataclass attribute (still) doesn't show up when hovering over usages of that attribute.

```python
class A(ModelBase):
    normal_attribute: int
    """normal docstring"""
    converted_attribute: int = model_field(converter=converter_simple)
    """converted docstring"""

a = A(normal_attribute=1, converted_attribute="2")
print(a.normal_attribute)
# Hovering on converted_attribute doesn't show docstring
print(a.converted_attribute)
```

![Image](https://github.com/user-attachments/assets/90df8024-6cd2-4686-be2d-db06cfb5a460)


To fix Pylance's semantic token provider, I used code like this:

```typescript
export function getDeclarationsForNameNodeWithSynthTypes(
    evaluator: TypeEvaluator,
    node: NameNode
): Declaration[] | undefined {
    const declInfo = evaluator.getDeclInfoForNameNode(node, /* skipUnreachableCode */ true);
    const declarations = declInfo?.decls;

    // If there are no declarations, but there is a synthesized type, then get the
    // declaration of the NameNode associated with the synthesized type instead.
    if ((!declarations || declarations.length === 0) && declInfo?.synthesizedTypes) {
        for (const synthType of declInfo.synthesizedTypes) {
            if (synthType.node) {
                const synthTypeDeclInfo = evaluator.getDeclInfoForNameNode(
                    synthType.node,
                    /* skipUnreachableCode */ true
                );

                return synthTypeDeclInfo?.decls;
            }
        }
    }

    return declarations;
}
```

I was able to use that same approach to fix the hover provider to show the docstring of `converted_attribute` above: https://github.com/debonte/pyright/commit/462ea82c2a455807210b013ddf2c680dbb245ddb

I found another spot in Pylance that can also be fixed this way.

Rather than reviewing every use of `getDeclInfoForNameNode` for synthesized type issues, would it make sense to add something like the code above into `getDeclInfoForNameNode`? Or are there possible negative side-effects of that?

**Comment by erictraut** [comment]
That doesn't look like the right approach to me. The `node` in a synthesized type should be used only for "go to" purposes. It doesn't have any additional meaning. It's not a real declaration, so it shouldn't be used to find doc strings, etc.

**Comment by debonte** [comment]
> That doesn't look like the right approach to me. The `node` in a synthesized type should be used only for "go to" purposes. It doesn't have any additional meaning. It's not a real declaration, so it shouldn't be used to find doc strings, etc.

Ok, how would you recommend getting the docstring in the dataclass converter scenario above?

Also, semantic tokenization isn't a "go to" scenario, but above you recommended using `SynthesizedTypeInfo.node` for that. So maybe it's useful outside of "go to purposes".

For what it's worth, I'm getting the `SynthesizedTypeInfo` from the declaration not the symbol. Not sure if that matters. While fixing the semantic token provider I realized that I could have done that before you made this fix.

**Comment by erictraut** [comment]
@debonte, you can add additional fields to the `SynthesizedTypeInfo` if that would be useful to the language service providers. The type evaluator wouldn't pay attention to these fields, so they wouldn't interfere with type checking functionality.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.386
